server/routerlicious/packages/services/src/documentManager.ts

```diff
@@ -5,21 +5,136 @@
+import {
+	IDocumentManager,
+	IDocument,
+	ITenantManager,
+	IDocumentStaticProperties,
+	ICache,
+} from "@fluidframework/server-services-core";
 import { generateToken, getCorrelationId } from "@fluidframework/server-services-utils";
+import * as Redis from "ioredis";
+import { Lumberjack, getLumberBaseProperties } from "@fluidframework/server-services-telemetry";
+import { RedisCache } from "./redis";
 
 /**
  * Manager to fetch document from Alfred using the internal URL.
  */
 export class DocumentManager implements IDocumentManager {
+	private static documentStaticDataCache: ICache;
+	/** True if the cache was initialized by setStaticProperty. Becomes false if the cache is overwritten by readDocument. */
+	private static staticCacheInitializedManually: boolean;
+
 	constructor(
 		private readonly internalAlfredUrl: string,
 		private readonly tenantManager: ITenantManager,
-	) {}
+	) {
+		// TODO: Eventually replace the static documentStaticDataCache field with a passed in optional value, to not have to initialize it here
+		// This can be done once the unified RedisCache/ICache implementation is made, the issue right now is that different RedisCache objects
+		// could be passed in, which would mean different headers and therefore multiple static document caches could exist at once.
+		// The unified RedisCache implementation will have more flexibility around headers, which will help fix this issue.
+
+		// If a redis cache does not yet exist, create one to populate the static field
+		if (!DocumentManager.documentStaticDataCache) {
+			const redisConfig: Redis.RedisOptions = {
+				host: "redis",
+				port: 6379,
+			};
+			const redisClient = new Redis.default(redisConfig);
+			DocumentManager.documentStaticDataCache = new RedisCache(redisClient);
+		}
+		if (!DocumentManager.staticCacheInitializedManually) {
+			DocumentManager.staticCacheInitializedManually = false;
+		}
+	}
 
 	public async readDocument(tenantId: string, documentId: string): Promise<IDocument> {
+		// Retrieve the document
 		const restWrapper = await this.getBasicRestWrapper(tenantId, documentId);
-		return restWrapper.get<IDocument>(`/documents/${tenantId}/${documentId}`);
+		const document: IDocument = await restWrapper.get<IDocument>(
+			`/documents/${tenantId}/${documentId}`,
+		);
+		if (!document) {
+			return undefined;
+		}
+		const staticProps: IDocumentStaticProperties =
+			DocumentManager.getStaticPropsFromDoc(document);
+
+		// Cache the static properties of the document
+		DocumentManager.staticCacheInitializedManually = false;
+		const staticPropsKey: string = DocumentManager.getDocumentStaticKey(documentId);
+		await DocumentManager.documentStaticDataCache.set(
+			staticPropsKey,
+			JSON.stringify(staticProps),
+		);
+
+		// Return the original document that was retrieved
+		return document;
+	}
+
+	public async readStaticProperty<K extends keyof IDocumentStaticProperties>(
```

Comment:
I think by building caching on property level, is over designed and very complex. Why not just read/set `IDocumentStaticProperties` as a whole like how `tenantCache` designed?

Reply:
The reason for this is so that the function can know which property we actually want from the static properties.
Knowing which property is being accessed allows for you to manually write to the cache, and then later know whether to overwrite the cache.

For example, for caching the ephemeral flag, we were previously creating an isEphemeral:<docId> cache key, and then later falling back to the cosmosDB if it is not there, and then internally the read document call is also checking its redis cache for static details. This means that isEphemeral property is being stored twice in the cache with that method. One possible thing to do here, is to just not create the isEphemeral:<docId> key, and rely only on readStaticProperties; however this leads to an extra cosmosDB call, since readDocument has not been called by the time the isEphemeral property is being checked.

To solve this issue, I allowed the user to manually set properties in the document static cache. This allows the part in historian with the isEphemeral flag caching to manually set the isEphemeralContainer property in the static cache. Since the cache has not been created at this point, doing this creates a new cache with all of the other static document props set to undefined, it also sets a flag in DocumentManager telling it that the cache was initialized manually. Later when some other part of the code tries to do readStaticProperty, we don't want it to grab an undefined value if it is trying to access something other than isEphemeral; however we also don't want  readStaticProperty to fallback to cosmosDB if we are just trying to read isEphemeral, because it is already cached there. The function knowing which property is being read allows it to decide whether the cache needs to be refreshed or not, if the cache was initialized manually and the property we are reading from it is undefined, we can assume that the property we want was never set and then fall back to cosmosDB.

Reply:
This is an example of how this could be used for ephemeral flag caching in Historian (not in this PR)
![image](https://github.com/microsoft/FluidFramework/assets/13680493/73817e9d-32f9-4cfb-82d5-e3cd8b9dc1cd)

Reply:
Had a chat with Pradeep about something similar to this. Will revert this to read/set IDocumentStaticProperties and remove the option for manually overriding properties. For the isEphemeralContainer caching workflow, we will just cache an isEphemeralContainer:<docId> key in redis. This will lead to some small duplicate data in redis, but shouldn't impact performance because it will avoid the cosmosDB call still, and in most cases will only require a single redis read operation.